pyteal/ast/scratch.py

```diff
@@ -59,40 +59,89 @@ def load(self, type: TealType = TealType.anytype) -> "ScratchLoad":
         """
         return ScratchLoad(self, type)
 
+    def index(self) -> "ScratchIndex":
+        return ScratchIndex(self)
+
     def __repr__(self):
         return "ScratchSlot({})".format(self.id)
 
     def __str__(self):
         return "slot#{}".format(self.id)
 
+
+ScratchSlot.__module__ = "pyteal"
+
+
+class ScratchIndex(Expr):
+    def __init__(self, slot: ScratchSlot):
+        super().__init__()
+        self.slot = slot
+
+    def __str__(self):
+        return "(ScratchIndex {})".format(self.slot)
+
     def __hash__(self):
-        return hash(self.id)
+        return hash(self.slot.id)
 
+    def type_of(self):
+        return TealType.uint64
 
-ScratchSlot.__module__ = "pyteal"
+    def has_return(self):
+        return False
+
+    def __teal__(self, options: "CompileOptions"):
+        from ..ir import TealOp, Op, TealBlock
+
+        op = TealOp(self, Op.int, self.slot)
+        return TealBlock.FromOp(options, op)
+
+
+ScratchIndex.__module__ = "pyteal"
 
 
 class ScratchLoad(Expr):
     """Expression to load a value from scratch space."""
 
-    def __init__(self, slot: ScratchSlot, type: TealType = TealType.anytype):
+    def __init__(
+        self,
+        slot: Optional[ScratchSlot],
+        type: TealType = TealType.anytype,
+        index_expression: Expr = None,
```

Comment:
Allowing the creation of `ScratchLoad` (and similarly for `ScratchStore`) by providing an expression for the index _instead_ of the currently required `slot`

Reply:
I held off commenting because I thought a backwards compatibility concern may explain the construction.

* Since there's latitude for API changes and because `slot` and `index_expression` are mutually exclusive, what do folks think of using `Union` here?
  * For example:  `slot: ScratchSlot | Expr`
* Aside:   Is it considered atypical to define a new `Expr` implementation specific to `Op.stores`?
  * Doing removes branching within `ScratchLoad`.
  * I don't have enough insight to understand if adding a new type relating to the scratch slot concept adds complexity elsewhere (e.g. IR).
  * In my _fuzzy_ mental model, I see a further extension.
    * _Disclaimer_:  These thoughts aren't fully formed.  And I don't think I understand the full implications.  I'm sharing to evidence my thinking.  It's _not_ a change request. 
    *  Define an `Expr` like `LeafExpr` that acts as the top-level type for any kind of scratch load.
    *  Define implementations specific to stack vs argument provided index values.
    * The approach enables abstracting over scratch loads independent of invocation scheme.
    * I can envision using pattern matching (as shown in https://docs.python.org/3/whatsnew/3.10.html#pep-634-structural-pattern-matching) to help group/isolate program behaviors.

Reply:
Good discussion point

Reply:
My initial thoughts are to stick with @jasonpaulos suggestions to maintain consistency and simplicity.

Reply:
I don't think we covered the topic verbally.

* Since I don't feel strongly and I don't want to hold up the PR, I'm on-board with @jasonpaulos's suggestion.
* If there's interest in the discussion point my prior remark raises, I'm happy to discuss live out-of-band to the PR.

Reply:
Gonna push back on this one, because optional arguments cannot come before required parameters so setting default `= None` would break the signature. So one has to explicitly provide `None` in the slot when an `index_expression` is provided.